app/models/report/standard_form_report.rb

```diff
@@ -89,16 +89,38 @@ def run(current_ability)
   end
 
   # returns all non-admin users in the form's mission with the given role that have not submitted any responses to the form
-  # options[:role] - (symbol) the role to check for
+  # options[:role]) the role to check for
+  # options[:limit]) how many users we want to fetch from the db
   def users_without_responses(options)
-    # eager load responses with users
-    all_observers = form.mission.assignments.includes(:user).find_all{|a| a.role.to_sym == options[:role] && !a.user.admin?}.map(&:user)
-    submitters = form.responses.includes(:user).map(&:user).uniq
-    @users_without_responses = all_observers - submitters
+    User.without_responses_for_form(form, options).sort
   end
 
-  def observers_without_responses
-    users_without_responses(:role => :observer)
+  # This returns the count of the previous users select without the limit clause.
+  # Format: [[count]]
+  def observers_without_responses_count
+    ActiveRecord::Base.connection.execute('SELECT FOUND_ROWS();').entries[0].first
```

Comment:
Great idea.

On Tue, Sep 15, 2015 at 1:17 PM, Leandro Matioli <notifications@github.com>
wrote:

> In app/models/report/standard_form_report.rb
> <https://github.com/thecartercenter/elmo/pull/141#discussion_r39537856>:
>
> >    end
> >
> > -  def observers_without_responses
> > -    users_without_responses(:role => :observer)
> > +  # This returns the count of the previous users select without the limit clause.
> > +  # Format: [[count]]
> > +  def observers_without_responses_count
> > +    ActiveRecord::Base.connection.execute('SELECT FOUND_ROWS();').entries[0].first
>
> Hum, i see your concern here... i separated them just to not have a method
> doing more stuff than it's supposed to (and to return user instances as
> it's expected).
>
> What do you think of returning a hash instead of an array? Something like:
> {users: users_found, count: users_count }. I think it's easier to use
> this when called, because you don't have to check on which position of the
> array the information you want is.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thecartercenter/elmo/pull/141/files#r39537856>.
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org